src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexEmbeddedCompletionProvider.cs

```diff
@@ -0,0 +1,560 @@
+        {
+            var replacementStart = parentOpt != null
+                ? parentOpt.GetSpan().Start
+                : context.Position;
+
+            var replacementSpan = TextSpan.FromBounds(replacementStart, context.Position);
+            var newPosition = replacementStart + positionOffset;
+
+            insertionText = insertionText ?? displayText;
+            var escapedInsertionText = _language.EscapeText(insertionText, stringToken);
+
+            if (escapedInsertionText != insertionText)
+            {
+                newPosition += escapedInsertionText.Length - insertionText.Length;
+            }
+
+            return new RegexItem(
+                displayText, suffix, description,
+                CompletionChange.Create(
+                    new TextChange(replacementSpan, escapedInsertionText),
+                    newPosition));
+        }
+
+        private (RegexNode parent, RegexToken Token)? FindToken(
+            RegexNode parent, VirtualChar ch)
+        {
+            foreach (var child in parent)
+            {
+                if (child.IsNode)
+                {
+                    var result = FindToken(child.Node, ch);
+                    if (result != null)
+                    {
+                        return result;
+                    }
+                }
+                else
+                {
+                    if (child.Token.VirtualChars.Contains(ch))
+                    {
+                        return (parent, child.Token);
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private bool IsInCharacterClass(RegexNode parent, VirtualChar ch, bool inCharacterClass)
+        {
+            foreach (var child in parent)
+            {
+                if (child.IsNode)
+                {
+                    var result = IsInCharacterClass(child.Node, ch, inCharacterClass || parent is RegexBaseCharacterClassNode);
+                    if (result)
+                    {
+                        return result;
+                    }
+                }
+                else
+                {
+                    if (child.Token.VirtualChars.Contains(ch))
+                    {
+                        return inCharacterClass;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        public override Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetValue(StartKey, out var startString) ||
+                !item.Properties.TryGetValue(LengthKey, out var lengthString) ||
+                !item.Properties.TryGetValue(NewTextKey, out var newText))
+            {
+                return SpecializedTasks.Default<CompletionChange>();
+            }
+
+            item.Properties.TryGetValue(NewPositionKey, out var newPositionString);
```

Comment:
When doesn't this have this? Not everything is going through `CreateItem` above?

Reply:
You're right.  will fix.

Reply:
Resolved.